homeassistant/components/apple_tv/config_flow.py

```diff
@@ -0,0 +1,438 @@
+    CONF_ADDRESS,
+    CONF_NAME,
+    CONF_PIN,
+    CONF_PROTOCOL,
+    CONF_TYPE,
+)
+from homeassistant.core import callback
+from homeassistant.data_entry_flow import AbortFlow
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import (
+    CONF_CREDENTIALS,
+    CONF_CREDENTIALS_AIRPLAY,
+    CONF_CREDENTIALS_DMAP,
+    CONF_CREDENTIALS_MRP,
+    CONF_IDENTIFIER,
+    CONF_START_OFF,
+)
+from .const import DOMAIN  # pylint: disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+
+INPUT_PIN_SCHEMA = vol.Schema({vol.Required(CONF_PIN, default=None): int})
+
+DEFAULT_START_OFF = False
+PROTOCOL_PRIORITY = [const.Protocol.MRP, const.Protocol.DMAP, const.Protocol.AirPlay]
+
+# Mapping between config entry format and pyatv
+CREDENTIAL_MAPPING = {
+    CONF_CREDENTIALS_MRP: const.Protocol.MRP.value,
+    CONF_CREDENTIALS_DMAP: const.Protocol.DMAP.value,
+    CONF_CREDENTIALS_AIRPLAY: const.Protocol.AirPlay.value,
+}
+
+
+async def device_scan(identifier, loop, cache=None):
+    """Scan for a specific device using identifier as filter."""
+
+    def _filter_device(dev):
+        if identifier is None:
+            return True
+        if identifier == str(dev.address):
+            return True
+        if identifier == dev.name:
+            return True
+        return any([service.identifier == identifier for service in dev.services])
+
+    def _host_filter():
+        try:
+            return [ip_address(identifier)]
+        except ValueError:
+            return None
+
+    if cache:
+        matches = [atv for atv in cache if _filter_device(atv)]
+        if matches:
+            return cache, matches[0]
+
+    for hosts in [_host_filter(), None]:
+        scan_result = await scan(loop, timeout=3, hosts=hosts)
+        matches = [atv for atv in scan_result if _filter_device(atv)]
+
+        if matches:
+            return scan_result, matches[0]
+
+    return scan_result, None
+
+
+def is_valid_credentials(credentials):
+    """Verify that credentials are valid for establishing a connection."""
+    return (
+        credentials.get(const.Protocol.MRP.value) is not None
+        or credentials.get(const.Protocol.DMAP.value) is not None
+    )
+
+
+class AppleTVConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Apple TV."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get options flow for this handler."""
+        return AppleTVOptionsFlow(config_entry)
+
+    def __init__(self):
+        """Initialize a new AppleTVConfigFlow."""
+        self.scan_result = None
+        self.atv = None
+        self.protocol = None
+        self.pairing = None
+        self.credentials = {}  # Protocol -> credentials
+
+    async def async_step_invalid_credentials(self, info):
```

Comment:
If it doesn't lead to user interaction then you don't need to call it a step

Reply:
I'm not sure how to refactor this in a good way. It's the flow framework that adds the step part (`async_setup` in `__init__.py`). I would still prefer to have the separation of concerns that I have, any ideas?

Reply:
Isn't invalid credentials just an ordinary method? It could be named whatever.

Reply:
Nah, it's the entry point for a custom flow source that I have defined (like user or zeroconf). It is started at row 263 in `config_flow.py` (not sure how to permalink within a PR).